### PR TITLE
Fix linker error of missing pthread on Linux

### PR DESCRIPTION
**Pull request by lordgamez**
Fixes error of issue #742

### CODE REVIEW

**Review by Jinming-Hu — APPROVED** [review]

File: sdk/core/azure-core/CMakeLists.txt [hunk 1]
```diff
@@ -8,9 +8,13 @@ project(${TARGET_NAME} LANGUAGES CXX)
 set(CMAKE_CXX_STANDARD 14)
 set(CMAKE_CXX_STANDARD_REQUIRED True)
 
+set(CMAKE_THREAD_PREFER_PTHREAD TRUE)
+set(THREADS_PREFER_PTHREAD_FLAG TRUE)
+find_package(Threads REQUIRED)
```
Comment (ahsonkhan) [inline sdk/core/azure-core/CMakeLists.txt:13]: Is this a linux only package? If so, should this be under a non-windows ifdef?

Comment (Jinming-Hu) [reply sdk/core/azure-core/CMakeLists.txt:13]: https://cmake.org/cmake/help/latest/module/FindThreads.html

It's **NOT** a linux only package. So it shouldn't under a non-windows ifdef.

File: sdk/core/azure-core/CMakeLists.txt [hunk 2]
```diff
@@ -58,7 +62,7 @@ add_library (Azure::Core ALIAS ${TARGET_NAME})
 create_code_coverage(core ${TARGET_NAME} "${TARGET_NAME}-test")
 
 target_include_directories(${TARGET_NAME} PUBLIC ${CURL_INCLUDE_DIRS})
-target_link_libraries(${TARGET_NAME} PRIVATE CURL::libcurl)
+target_link_libraries(${TARGET_NAME} PRIVATE CURL::libcurl Threads::Threads)
```
Comment (ahsonkhan) [inline sdk/core/azure-core/CMakeLists.txt:65]: `<thread>` is used in storage retry policy as well. Should we link `::Threads` there too, or is it sufficient to rely on azure-core linking it (which storage takes a dependency on)?

File: sdk/core/azure-core/CMakeLists.txt (hunk 1 above)
Comment (ahsonkhan) [reply sdk/core/azure-core/CMakeLists.txt:13]: I understand the `Threads` package isn't windows only. But what about the `pthread` related CMAKE flags. Those are posix only, right? What do `CMAKE_THREAD_PREFER_PTHREAD` and `THREADS_PREFER_PTHREAD_FLAG` being true mean on Windows? Are they ignored or will they look for pthread specifically and fail? Presumably on windows, the call to `find_package` will set `CMAKE_USE_WIN32_THREADS_INIT`.

File: sdk/core/azure-core/CMakeLists.txt (hunk 2 above)
Comment (ahsonkhan) [reply sdk/core/azure-core/CMakeLists.txt:65]: > Should we link `::Threads` there too, or is it sufficient to rely on azure-core linking it (which storage takes a dependency on)?

I just realized, storage common already links the `::Threads` library.

https://github.com/Azure/azure-sdk-for-cpp/blob/a716e094fe4c17f872181cc755e13843f53ead8d/sdk/storage/azure-storage-common/CMakeLists.txt#L44-L50

**Review by ahsonkhan — APPROVED** [review]

**Comment by ahsonkhan** [comment]
Thanks for the fix and contribution @lordgamez, and thank you @Jinming-Hu for the proposed solution.

File: sdk/core/azure-core/CMakeLists.txt (hunk 1 above)
Comment (Jinming-Hu) [reply sdk/core/azure-core/CMakeLists.txt:13]: whatever. It's a code change to core sdk. You guys decide what do with it.

**Pull request merged**